heat/core/vmap.py

```diff
@@ -0,0 +1,88 @@
+"""
+This implements a functionality similar to PyTorchs vmap function.
+"""
+
+import torch
+from .dndarray import DNDarray
+from .factories import array
+from .communication import MPI_WORLD
+
+__all__ = ["vmap"]
+
+
+def vmap(func, out_dims=0, randomness="error", *, chunk_size=None):
+    """
+    This function is used to apply a function to a DNDarray in a vectorized way.
+    `heat.vmap` return a callable that can be applied to DNDarrays.
+    Vectorization will automatically take place along the split axis/axes; therefore, unlike in PyTorch, there is no argument `in_dims`.
+
+    Parameters
+    ----------
+    func : callable
+        The function to apply to the DNDarray. It must take a DNDarray as its first argument.
+    out_dims : int or sequence of int, optional
+        The dimensions of the output(s) that are mapped over; identical to the split dimension(s) of the output(s).
+        Default is 0.
+    randomness : {'error', 'different', 'same'}, optional
+        Determines how to handle randomness in the function to be vmapped. This argument is directly passed to the underlying PyTorch vmaps;
+        see the corresponding PyTorch documentation for more information and the note below.
+        If 'error' (default), an error is raised if the function to be mapped contains randomness.
+    chunk_size : int, optional
+        The size of the chunks to use for the process-local computation.
+        If None (default), apply a single PyTorch vmap over the process-local chunks of data. If not None, then compute the process-local PyTorch vmap `chunk_size`
+        many samples at a time. Note that `chunk_size=1` is equivalent to computing the process-local PyTorch vmap's with a for-loop.
+        If you run into memory issues computing the vmap, please try a non-None chunk_size.
+
```

Comment:
Docstring and function declaration should include the signature of the returned function (if possible). Makes using vmap easier

Reply:
added signature with type hints